From ea62d3f86aef64ad08a10828ef66568b0b0ad753 Mon Sep 17 00:00:00 2001
From: John Levon <john.levon@joyent.com>
Date: Mon, 4 Mar 2019 14:11:56 +0000
Subject: [PATCH] OS-4950 bzip2 should be built with optimizations OS-7604
 -faggressive-loop-optimizations should be off everywhere Reviewed by: Robert
 Mustacchi <rm@joyent.com> Approved by: Jerry Jelinek
 <jerry.jelinek@joyent.com>

---
 Makefile.defs           | 13 ++++++++++++-
 bzip2/Makefile          | 10 +++-------
 bzip2/Makefile.com      | 23 +++--------------------
 bzip2/makefile.build    |  5 +----
 bzip2/makefile.build.64 | 32 --------------------------------
 perl/Makefile           |  2 +-
 uefi-edk2/Makefile      |  5 ++---
 7 files changed, 22 insertions(+), 68 deletions(-)
 delete mode 100644 bzip2/makefile.build.64

diff --git a/Makefile.defs b/Makefile.defs
index 2c37554..92537db 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -10,7 +10,7 @@
 #
 
 #
-# Copyright (c) 2018, Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 BASE =		$(PWD)
@@ -65,6 +65,17 @@ SEPARATE_BUILD =
 
 CPPFLAGS =	$(SYSINCDIRS:%=-isystem $(DESTDIR)/%)
 
+#
+# As we ship some rather downrev software, there's a good chance it might fall
+# foul of newer GCC's undefined-behavior optimizations; skip those. As it's in
+# general painful to ensure that $(CFLAGS) gets passed down correctly, we'll
+# just set it directly as the compiler...
+#
+ifneq ($(PRIMARY_COMPILER_VER),4)
+GCC += -fno-aggressive-loop-optimizations
+GXX += -fno-aggressive-loop-optimizations
+endif
+
 #
 # We ship platform-private perl modules (e.g. from ntp) in this directory,
 # and using this perl interpeter:
diff --git a/bzip2/Makefile b/bzip2/Makefile
index 995acc1..2c95713 100644
--- a/bzip2/Makefile
+++ b/bzip2/Makefile
@@ -56,10 +56,6 @@ $(VER)/Makefile.dist: $(BVER).tar.gz
 	(cd $(VER); gpatch -p1 < ../bzip2.patch)
 	cp $(VER)/Makefile $(VER)/Makefile.dist
 
-$(VER)/i386/Makefile: $(VER)/Makefile.dist makefile.build
-	mkdir -p $(VER)/i386
-	cp makefile.build $(VER)/i386/Makefile
-
-$(VER)/amd64/Makefile: $(VER)/Makefile.dist makefile.build.64
-	mkdir -p $(VER)/amd64
-	cp makefile.build.64 $(VER)/amd64/Makefile
+$(VER)/i386/Makefile $(VER)/amd64/Makefile: $(VER)/Makefile.dist makefile.build
+	mkdir -p $(@D)
+	cp makefile.build $@
diff --git a/bzip2/Makefile.com b/bzip2/Makefile.com
index f10099b..d24f915 100644
--- a/bzip2/Makefile.com
+++ b/bzip2/Makefile.com
@@ -21,7 +21,7 @@
 # Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
 #
-# Copyright (c) 2012 Joyent Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 LIBRARY = libbz2.a
@@ -51,7 +51,7 @@ endif
 LDLIBS += -lc
 
 PROGS = bzip2 bzip2recover
-PROG_LDFLAGS += -Wl,-Bdirect -L. 
+PROG_LDFLAGS += -Wl,-Bdirect -L.
 ifneq ($(STRAP),strap)
 	PROG_LDFLAGS += $(GENLDFLAGS)
 endif
@@ -59,9 +59,8 @@ bzip2_LDLIBS += -lbz2
 
 # --- Generic ---
 
-CC = gcc
 CPPFLAGS += -I..
-CFLAGS += -fPIC
+CFLAGS += -fPIC -O2
 DYNLIB = $(LIBRARY:.a=.so)$(VERS)
 LIBLINKS = $(LIBRARY:.a=.so)
 
@@ -87,19 +86,3 @@ bzip2: $(LIBLINKS) $(DYNLIB) $(bzip2_OBJECTS)
 
 bzip2recover: $(bzip2recover_OBJECTS)
 	$(LINK.prog)
-
-test: $(PROGS)
-	@cat ../words1
-	env LD_LIBRARY_PATH=`pwd` ./bzip2 -1  < ../sample1.ref > sample1.rb2
-	env LD_LIBRARY_PATH=`pwd` ./bzip2 -2  < ../sample2.ref > sample2.rb2
-	env LD_LIBRARY_PATH=`pwd` ./bzip2 -3  < ../sample3.ref > sample3.rb2
-	env LD_LIBRARY_PATH=`pwd` ./bzip2 -d  < ../sample1.bz2 > sample1.tst
-	env LD_LIBRARY_PATH=`pwd` ./bzip2 -d  < ../sample2.bz2 > sample2.tst
-	env LD_LIBRARY_PATH=`pwd` ./bzip2 -ds < ../sample3.bz2 > sample3.tst
-	cmp ../sample1.bz2 sample1.rb2 
-	cmp ../sample2.bz2 sample2.rb2
-	cmp ../sample3.bz2 sample3.rb2
-	cmp sample1.tst ../sample1.ref
-	cmp sample2.tst ../sample2.ref
-	cmp sample3.tst ../sample3.ref
-	@cat ../words3
diff --git a/bzip2/makefile.build b/bzip2/makefile.build
index b09f53b..8c6eb69 100644
--- a/bzip2/makefile.build
+++ b/bzip2/makefile.build
@@ -21,12 +21,9 @@
 # Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
 # Use is subject to license terms.
 #
-# Copyright (c) 2012 Joyent Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 CPPFLAGS = -D_FILE_OFFSET_BITS=64
-CFLAGS = -m32
-LDFLAGS = -m32
-PROG_LDFLAGS = -m32
 
 include ../../Makefile.com
diff --git a/bzip2/makefile.build.64 b/bzip2/makefile.build.64
deleted file mode 100644
index 107db86..0000000
--- a/bzip2/makefile.build.64
+++ /dev/null
@@ -1,32 +0,0 @@
-#
-#
-# CDDL HEADER START
-#
-# The contents of this file are subject to the terms of the
-# Common Development and Distribution License (the "License").
-# You may not use this file except in compliance with the License.
-#
-# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
-# or http://www.opensolaris.org/os/licensing.
-# See the License for the specific language governing permissions
-# and limitations under the License.
-#
-# When distributing Covered Code, include this CDDL HEADER in each
-# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
-# If applicable, add the following below this CDDL HEADER, with the
-# fields enclosed by brackets "[]" replaced with your own identifying
-# information: Portions Copyright [yyyy] [name of copyright owner]
-#
-# CDDL HEADER END
-#
-# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
-# Use is subject to license terms.
-#
-# Copyright (c) 2012 Joyent Inc.
-#
-
-CFLAGS = -m64
-LDFLAGS = -m64
-PROG_LDFLAGS = -m64
-
-include ../../Makefile.com
diff --git a/perl/Makefile b/perl/Makefile
index 3d3b938..c8100cc 100644
--- a/perl/Makefile
+++ b/perl/Makefile
@@ -149,7 +149,7 @@ $(CONFIG_PL): $(CONFIG_PL).in Makefile
 
 $(VER.32)/cflags: $(VER.32)/config.over $(VER.32)/Configure
 	(cd $(VER.32) && \
-	    ./Configure -des -Dcc=$(GCC) -Duse64bitint)
+	    ./Configure -des -Dcc="$(GCC)" -Duse64bitint)
 
 $(VER.32)/perldtrace.h: $(VER.32)/cflags
 	(cd $(VER.32) && \
diff --git a/uefi-edk2/Makefile b/uefi-edk2/Makefile
index 9d5bee3..b658499 100644
--- a/uefi-edk2/Makefile
+++ b/uefi-edk2/Makefile
@@ -10,7 +10,7 @@
 #
 
 #
-# Copyright 2018 Joyent, Inc.
+# Copyright (c) 2019, Joyent, Inc.
 #
 
 
@@ -28,8 +28,7 @@ MAKE_ARGS+=	AS=$(STRAPPROTO)/usr/gnu/bin/gas \
 		AR=$(STRAPPROTO)/usr/gnu/bin/gar \
 		LD=$(STRAPPROTO)/usr/gnu/bin/gld \
 		OBJCOPY=$(STRAPPROTO)/usr/gnu/bin/gobjcopy \
-		CC=$(STRAPPROTO)/usr/bin/gcc \
-		CXX=$(STRAPPROTO)/usr/bin/g++
+		CC="$(CC)" CXX="$(CXX)"
 
 ARCH=		X64
 UEFI_TARGET=	RELEASE
-- 
2.21.0

